languageFactories/language-factory-core/src/main/scala/cromwell/languages/util/ImportResolver.scala

```diff
@@ -12,86 +13,151 @@ import com.softwaremill.sttp.asynchttpclient.cats.AsyncHttpClientCatsBackend
-    */
-  def httpResolver: ImportResolver =
-    httpResolverWithHeaders(Map.empty[String,String])
+  case class HttpResolver(relativeTo: Option[String] = None, headers: Map[String, String] = Map.empty) extends ImportResolver {
+    import HttpResolver._
 
-  /**
-    * Given a string representing an http(s) url, retrieve the contents
-    * using the supplied headers (which can be an empty map)
-    */
-  def httpResolverWithHeaders(headers: Map[String,String]): ImportResolver = CheckedAtoB.fromCheck { str: String =>
+    override def name: String = {
+      s"http importer${relativeTo map { r => s" (relative to $r)" } getOrElse ""}"
+    }
+
+    def newResolverList(newRoot: String): List[ImportResolver] = {
+      val rootWithoutFilename = newRoot.split('/').init.mkString("", "/", "/")
+      List(
+        HttpResolver(relativeTo = Some(canonicalize(rootWithoutFilename)), headers)
+      )
+    }
 
-    implicit val sttpBackend = AsyncHttpClientCatsBackend[IO]()
+    def pathToLookup(str: String): Checked[String] = relativeTo match {
+      case Some(relativeToValue) =>
+        if (str.startsWith("http")) canonicalize(str).validNelCheck
+        else if (str.startsWith("/")) canonicalize(s"${root(relativeToValue).stripSuffix("/")}/$str").validNelCheck
+        else canonicalize(s"${relativeToValue.stripSuffix("/")}/$str").validNelCheck
+      case None =>
+        if (str.startsWith("http")) canonicalize(str).validNelCheck
+        else s"Cannot import '$str' relative to nothing".invalidNelCheck
+    }
 
-    val responseIO : IO[Response[String]] =
-      sttp.get(uri"$str").headers(headers).send()
+    override def innerResolver(str: String, currentResolvers: List[ImportResolver]): Checked[ResolvedImportBundle] = {
+      pathToLookup(str) flatMap { toLookup =>
+        (Try {
+          implicit val sttpBackend = AsyncHttpClientCatsBackend[IO]()
+          val responseIO: IO[Response[String]] = sttp.get(uri"$toLookup").headers(headers).send()
+
+          // temporary situation to get functionality working before
+          // starting in on async-ifying the entire WdlNamespace flow
+          val result: Checked[String] = Await.result(responseIO.unsafeToFuture, 15.seconds).body.leftMap(NonEmptyList(_, List.empty))
+
+          result map {
+            ResolvedImportBundle(_, newResolverList(toLookup))
+          }
+        } match {
+          case Success(result) => result
+          case Failure(e) => s"HTTP resolver with headers had an unexpected error (${e.getMessage})".invalidNelCheck
+        }).contextualizeErrors(s"download $toLookup")
+      }
+    }
+  }
 
-    // temporary situation to get functionality working before
-    // starting in on async-ifying the entire WdlNamespace flow
-    val result: Checked[String] = Await.result(responseIO.unsafeToFuture, 15.seconds).body.leftMap(NonEmptyList(_, List.empty))
+  object HttpResolver {
+    def canonicalize(str: String): String = {
```

Comment:
FWIW, I had to look up what canonical means when applied to filepaths

Perhaps I am the last person on earth who didn't know, or perhaps this could use a brief comment

I like this explanation...
>For example, both ../../here/bar/x and ./test/../../bar/x may refer to the same location, but you can't do a textual comparison on the two paths. However, if you turn them into their canonical representation, they both become ../bar/x, and we see that they actually refer to the same thing.

...because it makes clear how they differ from absolute paths (which would also allow textual comparison)

[Source.](https://stackoverflow.com/a/12100371/818054)

Reply:
To be fair, it **is** part of the [Java `File` API](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#getCanonicalFile())